### PR TITLE
[FSDP][optim_state_dict] Consolidate rank0_only load logic

### PR DESCRIPTION
**Pull request by fegin**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #99731
* __->__ #99647
* #99624

Follow up https://github.com/pytorch/pytorch/pull/99624, this PR consolidate the logic of `use_orig_params=False` with `use_orig_params=True` to use the same logic to load optimizer checkpoint when rank0_only is True.

### CODE REVIEW

File: torch/distributed/fsdp/_optim_utils.py [hunk 1]
```diff
@@ -493,6 +490,7 @@ def _flatten_optim_state_dict(
                     else:
                         # Move the tensor in the original osd back to CPU to make the
                         # original osd unaffected.
+                        continue
```
Comment (wz337) [inline torch/distributed/fsdp/_optim_utils.py:493]: QQ about the continue: Are we just skipping moving the state back to CPU here?

**Review by wz337 — APPROVED** [review]
LGTM!

File: torch/distributed/fsdp/_optim_utils.py (hunk 1 above)
Comment (fegin) [reply torch/distributed/fsdp/_optim_utils.py:493]: Oop, sorry, didn't remove this before submitting the PR. Will update it.

**Pull request closed without merging**